extensions/collection.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2024 Contributors to the Veraison project.
+// SPDX-License-Identifier: Apache-2.0
+
+package extensions
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	cbor "github.com/fxamacker/cbor/v2"
+)
+
+var ErrUnexpectedPoint = errors.New("unexpected extension point")
+
+// Point defines an extension point -- a place where a struct defining
+// extensions may be attached. There is at least one Point corresponding to
+// each struct that embeds Extensions.
+type Point string
+
+// Map is a mapping of extension Point's to IMapValue's (pointers to structs that
+// define extension fields.
+type Map map[Point]IMapValue
+
+// NewMap instantiates an empty Map
+func NewMap() Map {
+	return make(Map)
+}
+
+// Add a Paint-IMapValue entry to the map. If there is already an IMapValue
+// associated with the Point, it will be replaced.
+func (o Map) Add(p Point, v IMapValue) Map {
+	o[p] = v
+	return o
+}
+
+// IExtensible defines an interface for extensible objects (those that can
+// register extensions.
+type IExtensible[P any] interface {
+	RegisterExtensions(exts Map) error
+	GetExtensions() IMapValue
+	Valid() error
+
+	*P // this interface must be implemented by a pointer to P
+}
+
+// Collection is a generic container for objects who's pointers implement
+// IExtensible. In addition of containing instances of extensible objects, it
+// is capable of registering extensions for those instances.
+type Collection[P any, I IExtensible[P]] struct {
+	Values []P
+
+	valueExtensions cache
+}
+
+// NewCollection instantiates an empty Measurements array
```

Comment:
Is the comment on line, correct ? I thought it is generic to any extensible array?

Reply:
Argh, yeah, I started out doing this for measurements, and then realised that I'd need to do that for every extensible type, so made it generic. the comment is left over from old implmentation; will fix.